src/shogun/machine/GLM.cpp

```diff
@@ -0,0 +1,408 @@
+	}
+}
+
+void GLM::iteration()
+{
+	std::shared_ptr<ElasticNetPenalty> penalty;
+	penalty->set_l1_ratio(m_alpha);
+
+	SGMatrix<float64_t> X = LinearMachine::features->get_computed_dot_feature_matrix();
+	SGVector<float64_t> y = m_labels->get_values();
+	SGVector<float64_t> w_old(m_w);
+	SGVector<float64_t> gradient_w = compute_grad_L2_loss_w(X, y, m_w, bias);
+	float64_t gradient_bias = compute_grad_L2_loss_bias(X, y, m_w, bias);
+
+	for (int i = 0; i < m_w.vlen; i++)
+	{
+		m_w[i] += m_lambda * penalty->get_penalty_gradient(m_w[i], 0.0);
+	}
+
+	//Update
+	m_w = linalg::add(m_w, gradient_w, 1.0, -1*m_learning_rate);
+	if(m_compute_bias)
+		bias = bias - m_learning_rate * gradient_bias;
+	
+	//Apply proximal operator
+	penalty->update_variable_for_proximity(m_w, m_lambda * m_alpha);
+
+	//Convergence by relative parameter change tolerance
+	float64_t norm_update = linalg::norm(linalg::add(m_w, w_old, 1.0, -1.0));
+	if(m_current_iteration > 0 && (norm_update / linalg::norm(m_w)) < m_tolerance)
+		m_complete = true;
+}
+
+const SGVector<float64_t> GLM::conditional_intensity(const SGMatrix<float64_t> X, const SGVector<float64_t> w,const float64_t l_bias)
+{
+	SGVector<float64_t> z = compute_z(X, w, l_bias);
+	SGVector<float64_t> result = non_linearity(z);
+	return result;
+}
+
+const SGVector<float64_t> GLM::compute_z(const SGMatrix<float64_t> X, const SGVector<float64_t> w, const float64_t l_bias)
+{
+	SGVector<float64_t> z = linalg::matrix_prod(X, w, false);
+	return z;
+}
+
+const SGVector<float64_t> GLM::non_linearity(const SGVector<float64_t> z)
+{
+	SGVector<float64_t> result;
+	switch (distribution)
+	{
+	case POISSON:
+		result = SGVector<float64_t>(z);
+		float64_t l_bias = 0;
+
+		if(LinearMachine::m_compute_bias)
+			l_bias = (1 - m_eta) * std::exp(m_eta);
+
+		for (int i = 0; i < z.vlen; i++)
+		{
+			if(z[i]>m_eta)
+				result[i] = z[i] * std::exp(m_eta) + l_bias;
+			else
+				result[i] = std::exp(z[i]);
+		}
+		break;
+	
+	default:
+		error("Not a valid distribution type");
+		break;
+	}
+	return result;
+}
+
+const SGVector<float64_t> GLM::gradient_non_linearity(const SGVector<float64_t> z)
+{
+	SGVector<float64_t> result;
+	switch (distribution)
+	{
+	case POISSON:
+		result = SGVector<float64_t>(z);
+		for (int i = 0; i < z.vlen; i++)
+		{
+			if(z[i]>m_eta)
+				
+				result[i] = std::exp(m_eta);
+			else
+				result[i] = std::exp(z[i]);
+		}
+		break;
+	
+	default:
+		error("Not a valid distribution type");
+		break;
+	}
+	return result;
+}
+
+const SGVector<float64_t> GLM::compute_grad_L2_loss_w(const SGMatrix<float64_t> X, const SGVector<float64_t> y, const SGVector<float64_t> w, const float64_t l_bias)
+{
+	auto n_samples = y.vlen;
+	auto n_features = X.num_rows;
+	
+	SGVector<float64_t> z = compute_z(X, w, l_bias);
+	SGVector<float64_t> mu = non_linearity(z);
+	SGVector<float64_t> grad_mu = gradient_non_linearity(z);
+
+	SGVector<float64_t> grad_w(w.vlen);
+	switch (distribution)
+	{
+	case POISSON:
```

Comment:
could this code pls leave in another class/structure. A local one is fine, but we want to shorten these type of methods

Reply:
Yes I have done something like that when I implement the SGDMinimizer class
I will remove all this code once you approve.

Reply:
So, Should I move these methods to another class and use the `IterativeMachine`?
or
Should I call the `SGDMinimizer` (one of the optimization classes) class and use that to minimize the cost funtion I have defined in the `GLMCostFunction` class

Reply:
Yes I'll do this